File: client/ios/BVGMessage.swift
import Foundation

struct Intent: Codable {
    let primary: String
    let confidence: Double?
    
    init(primary: String, confidence: Double = 1.0) {
        self.primary = primary
        self.confidence = confidence
    }
}

struct State: Codable {
    let timestamp: Int64
    let sender: String
    let receiver: String
    let status: String
    
    init(timestamp: Int64, sender: String, receiver: String, status: String = "sent") {
        self.timestamp = timestamp
        self.sender = sender
        self.receiver = receiver
        self.status = status
    }
}

struct Content: Codable {
    let text: String?
    let media: Data?
    
    init(text: String? = nil, media: Data? = nil) {
        self.text = text
        self.media = media
    }
}

struct Meta: Codable {
    let intent: Intent
    let state: State
}

struct BVGMessage: Codable {
    let id: String
    let type: String
    let content: Content
    let meta: Meta
    
    init(id: String = UUID().uuidString,
         type: String = "message",
         content: Content,
         meta: Meta) {
        self.id = id
        self.type = type
        self.content = content
        self.meta = meta
    }
    
    func serialize() -> Data? {
        let encoder = JSONEncoder()
        return try? encoder.encode(self)
    }
    
    static func deserialize(_ data: Data) -> BVGMessage? {
        let decoder = JSONDecoder()
        return try? decoder.decode(BVGMessage.self, from: data)
    }
    
    static func createTextMessage(text: String, from: String, to: String) -> BVGMessage {
        return BVGMessage(
            content: Content(text: text),
            meta: Meta(
                intent: Intent(primary: "chat"),
                state: State(
                    timestamp: Int64(Date().timeIntervalSince1970 * 1000),
                    sender: from,
                    receiver: to
                )
            )
        )
    }
}

File: client/android/BvgMessage.kt
package com.bridge.messenger

import kotlinx.serialization.*
import kotlinx.serialization.json.*
import java.util.*

@Serializable
data class Intent(
    val primary: String,
    val confidence: Float = 1.0f
)

@Serializable
data class State(
    val timestamp: Long,
    val sender: String,
    val receiver: String,
    val status: String = "sent" // sent, delivered, read
)

@Serializable
data class Content(
    val text: String? = null,
    val media: ByteArray? = null
)

@Serializable
data class BVGMessage(
    val id: String,
    val type: String = "message",
    val content: Content,
    val meta: Meta
) {
    @Serializable
    data class Meta(
        val intent: Intent,
        val state: State
    )

    fun serialize(): ByteArray {
        return Json.encodeToString(this).toByteArray()
    }

    companion object {
        fun deserialize(data: ByteArray): BVGMessage {
            val json = String(data)
            return Json.decodeFromString(json)
        }

        fun createTextMessage(text: String, from: String, to: String): BVGMessage {
            return BVGMessage(
                id = UUID.randomUUID().toString(),
                content = Content(text = text),
                meta = Meta(
                    intent = Intent(primary = "chat"),
                    state = State(
                        timestamp = System.currentTimeMillis(),
                        sender = from,
                        receiver = to
                    )
                )
            )
        }
    }
}